File: app/mobile/lib/features/channels/channel_kinds.dart
// Mobile mirror of app/shared/src/lib/channelKinds.ts. Drives the
// kind-picker FAB sheet, the create/edit form, and the per-row
// metadata (which field to mask as a "token preview"). Bridge is
// intentionally absent — its create flow needs a token generator
// and capability multiselect that don't translate cleanly to the
// mobile form pattern; bridge channels stay web-only on creation.

import 'package:opendray/core/i18n/strings.g.dart';

enum ChannelFieldType { text, password, textarea, boolean }

class ChannelField {
  const ChannelField({
    required this.name,
    required this.label,
    required this.type,
    this.required = false,
    this.placeholder,
    this.hint,
    this.optional = false,
    this.defaultValue,
  });

  final String name;
  final String label;
  final ChannelFieldType type;
  final bool required;
  final String? placeholder;
  final String? hint;
  // optional = true means: when the field is left blank on create,
  // omit it from the submitted config so server defaults apply
  // (rather than persisting an empty string).
  final bool optional;
  // Initial value for `boolean` fields when the channel config has no
  // stored value. (`default` is a Dart reserved word, hence the name.)
  final bool? defaultValue;
}

class ChannelKind {
  const ChannelKind({
    required this.kind,
    required this.label,
    required this.description,
    required this.fields,
    this.tokenFields = const [],
    this.webhookBased = false,
    this.afterCreateHint,
  });

  final String kind;
  final String label;
  final String description;
  final List<ChannelField> fields;
  // Fields whose presence in `config` is used as the masked "token"
  // preview on the row. First non-empty wins.
  final List<String> tokenFields;
  // When true the kind needs the operator to paste a publicly-
  // reachable webhook URL into the platform's admin console; the
  // post-create dialog surfaces the URL with a copy button.
  final bool webhookBased;
  // Optional callout shown after a successful create.
  final String? afterCreateHint;
}

// Runtime builder — i18n strings resolved against the active locale
// each time it's called. Cheap (a few dozen string lookups) and the
// callers already invoke it once per rebuild, so memoizing isn't
// worth the staleness risk.
List<ChannelKind> channelKindsList() => [
      ChannelKind(
        kind: 'telegram',
        label: 'Telegram',
        description: t.channels.kinds.telegram.description,
        tokenFields: const ['bot_token'],
        fields: [
          ChannelField(
            name: 'bot_token',
            label: t.channels.kinds.telegram.botTokenLabel,
            type: ChannelFieldType.password,
            required: true,
            placeholder: '123456:ABC-DEF...',
            hint: t.channels.kinds.telegram.botTokenHint,
          ),
          ChannelField(
            name: 'chat_id',
            label: t.channels.kinds.telegram.chatIdLabel,
            type: ChannelFieldType.text,
            placeholder: t.channels.kinds.telegram.chatIdPlaceholder,
            optional: true,
          ),
          ChannelField(
            name: 'owner_user_ids',
            label: t.channels.kinds.telegram.ownerUserIdsLabel,
            type: ChannelFieldType.text,
            optional: true,
            placeholder: t.channels.kinds.telegram.ownerUserIdsPlaceholder,
            hint: t.channels.kinds.telegram.ownerUserIdsHint,
          ),
          ChannelField(
            name: 'chat_enabled',
            label: t.channels.kinds.telegram.chatEnabledLabel,
            type: ChannelFieldType.boolean,
            defaultValue: true,
            hint: t.channels.kinds.telegram.chatEnabledHint,
          ),
          ChannelField(
            name: 'chat_typing',
            label: t.channels.kinds.telegram.chatTypingLabel,
            type: ChannelFieldType.boolean,
            defaultValue: true,
            hint: t.channels.kinds.telegram.chatTypingHint,
          ),
          ChannelField(
            name: 'reply_max_chars',
            label: t.channels.kinds.telegram.replyMaxCharsLabel,
            type: ChannelFieldType.text,
            optional: true,
            placeholder: t.channels.kinds.telegram.replyMaxCharsPlaceholder,
            hint: t.channels.kinds.telegram.replyMaxCharsHint,
          ),
        ],
      ),
      ChannelKind(
        kind: 'slack',
        label: 'Slack',
        description: t.channels.kinds.slack.description,
        tokenFields: const ['bot_token'],
        fields: [
          ChannelField(
            name: 'bot_token',
            label: t.channels.kinds.slack.botTokenLabel,
            type: ChannelFieldType.password,
            required: true,
            placeholder: 'xoxb-...',
            hint: t.channels.kinds.slack.botTokenHint,
          ),
          ChannelField(
            name: 'app_token',
            label: t.channels.kinds.slack.appTokenLabel,
            type: ChannelFieldType.password,
            required: true,
            placeholder: 'xapp-...',
            hint: t.channels.kinds.slack.appTokenHint,
          ),
          ChannelField(
            name: 'channel_id',
            label: t.channels.kinds.slack.channelIdLabel,
            type: ChannelFieldType.text,
            placeholder: t.channels.kinds.slack.channelIdPlaceholder,
            optional: true,
          ),
        ],
      ),
      ChannelKind(
        kind: 'discord',
        label: 'Discord',
        description: t.channels.kinds.discord.description,
        tokenFields: const ['bot_token'],
        fields: [
          ChannelField(
            name: 'bot_token',
            label: t.channels.kinds.discord.botTokenLabel,
            type: ChannelFieldType.password,
            required: true,
            placeholder: t.channels.kinds.discord.botTokenPlaceholder,
            hint: t.channels.kinds.discord.botTokenHint,
          ),
          ChannelField(
            name: 'channel_id',
            label: t.channels.kinds.discord.channelIdLabel,
            type: ChannelFieldType.text,
            placeholder: t.channels.kinds.discord.channelIdPlaceholder,
            optional: true,
          ),
        ],
      ),
      ChannelKind(
        kind: 'feishu',
        label: 'Feishu (飞书)',
        description: t.channels.kinds.feishu.description,
        tokenFields: const ['app_secret'],
        webhookBased: true,
        afterCreateHint: t.channels.kinds.feishu.afterCreateHint,
        fields: [
          ChannelField(
            name: 'app_id',
            label: t.channels.kinds.feishu.appIdLabel,
            type: ChannelFieldType.text,
            required: true,
            placeholder: 'cli_a1b2c3d4...',
          ),
          ChannelField(
            name: 'app_secret',
            label: t.channels.kinds.feishu.appSecretLabel,
            type: ChannelFieldType.password,
            required: true,
            placeholder: t.channels.kinds.feishu.appSecretPlaceholder,
          ),
          ChannelField(
            name: 'verification_token',
            label: t.channels.kinds.feishu.verificationTokenLabel,
            type: ChannelFieldType.password,
            optional: true,
            hint: t.channels.kinds.feishu.verificationTokenHint,
          ),
          ChannelField(
            name: 'chat_id',
            label: t.channels.kinds.feishu.chatIdLabel,
            type: ChannelFieldType.text,
            placeholder: t.channels.kinds.feishu.chatIdPlaceholder,
            optional: true,
          ),
        ],
      ),
      ChannelKind(
        kind: 'dingtalk',
        label: 'DingTalk (钉钉)',
        description: t.channels.kinds.dingtalk.description,
        tokenFields: const ['secret', 'webhook_url'],
        fields: [
          ChannelField(
            name: 'webhook_url',
            label: t.channels.kinds.dingtalk.webhookUrlLabel,
            type: ChannelFieldType.password,
            required: true,
            placeholder: 'https://oapi.dingtalk.com/robot/send?access_token=...',
          ),
          ChannelField(
            name: 'secret',
            label: t.channels.kinds.dingtalk.secretLabel,
            type: ChannelFieldType.password,
            optional: true,
            placeholder: 'SEC...',
            hint: t.channels.kinds.dingtalk.secretHint,
          ),
        ],
      ),
      ChannelKind(
        kind: 'wecom',
        label: 'WeCom (企业微信)',
        description: t.channels.kinds.wecom.description,
        tokenFields: const ['webhook_key', 'webhook_url'],
        fields: [
          ChannelField(
            name: 'webhook_key',
            label: t.channels.kinds.wecom.webhookKeyLabel,
            type: ChannelFieldType.password,
            required: true,
            placeholder: t.channels.kinds.wecom.webhookKeyPlaceholder,
            hint: t.channels.kinds.wecom.webhookKeyHint,
          ),
          ChannelField(
            name: 'webhook_url',
            label: t.channels.kinds.wecom.webhookUrlLabel,
            type: ChannelFieldType.password,
            optional: true,
            placeholder: 'https://qyapi.weixin.qq.com/...',
          ),
        ],
      ),
      // Personal-WeChat (WxPusher) entry removed from the create
      // flow — see app/shared/src/lib/channelKinds.ts for the
      // rationale. The server adapter remains for back-compat.
    ];

ChannelKind? findKind(String kind) {
  for (final k in channelKindsList()) {
    if (k.kind == kind) return k;
  }
  return null;
}

// Reads the masked token preview value from a config map for a kind.
// Returns the full string — caller decides whether to mask.
String? extractTokenPreview(ChannelKind kind, Map<String, dynamic> config) {
  for (final f in kind.tokenFields) {
    final v = config[f];
    if (v is String && v.isNotEmpty) return v;
  }
  return null;
}

String maskToken(String raw) {
  if (raw.length <= 10) return '••••';
  return '${raw.substring(0, 6)}…${raw.substring(raw.length - 4)}';
}

/// Serialize a channel-config field's (already-trimmed) text value to the
/// JSON type the gateway expects, mirroring the web form
/// (app/web/src/pages/Channels.tsx → buildConfigFromValues).
///
/// Only `chat_id` needs coercion among the kinds the mobile form
/// supports: Telegram persists it as an `int64`, so submitting it as a
/// string makes the server reject the whole config —
/// "json: cannot unmarshal string into Go struct field config.chat_id
/// of type int64" — which is why a Telegram channel configured from the
/// phone failed to start. The numeric guard leaves every other value a
/// string, exactly as the web form submits it:
///   - Feishu's `oc_…` chat_id (non-numeric, stays a string),
///   - Slack/Discord `channel_id` snowflakes (18-digit IDs that must NOT
///     be narrowed to a number — precision loss; note this is a
///     different field name, so it is never matched here anyway),
///   - `owner_user_ids` / `reply_max_chars` / tokens (the server reads
///     reply_max_chars from a number or a numeric string, so a string is
///     fine).
///
/// The mobile form exposes no `topic_ids` / `uids` fields, so — unlike
/// the web builder — those need no handling here.
Object coerceChannelConfigValue(String name, String raw) {
  if (name == 'chat_id' && _bareIntPattern.hasMatch(raw)) {
    return int.tryParse(raw) ?? raw;
  }
  return raw;
}

// A bare integer with an optional leading '-' (Telegram group /
// supergroup ids look like -1001234567890). Mirrors the web guard
// /^-?\d+$/ so numeric-looking-but-not-numeric values stay strings.
final RegExp _bareIntPattern = RegExp(r'^-?\d+$');

/// Render a stored config value as the text its TextField should show
/// when editing — the inverse of [coerceChannelConfigValue]. A numeric
/// chat_id (persisted as a JSON number) must round-trip back to its
/// digits; without this the edit form seeds the field empty and the next
/// save silently drops the chat_id (losing outbound notifications).
/// Absent / non-scalar values seed empty.
String channelConfigFieldText(Object? value) {
  if (value is String) return value;
  if (value is num) return value.toString();
  return '';
}

File: app/shared/src/lib/channelKinds.ts
// Catalogue of all channel kinds opendray ships natively. Drives the
// admin UI's New Channel dialog (form fields, hints, defaults) and the
// per-card display (which credential field to mask and show as
// "token: 1234…abcd").
//
// `bridge` is intentionally absent — it has its own dedicated UI
// (token generator, capability multiselect, post-create adapter
// setup dialog) wired in Channels.tsx.

export type FieldType = 'text' | 'password' | 'textarea' | 'boolean'

export interface KindField {
  name: string
  label: string
  type: FieldType
  required?: boolean
  placeholder?: string
  hint?: string
  /** When true and not provided by the user, the field is omitted from
   * the submitted config (so server defaults apply). */
  optional?: boolean
  /** Initial value for `boolean` fields when the channel config has no
   * stored value (new channel, or a field added after the channel was
   * created). Ignored for non-boolean fields. */
  default?: boolean
}

export interface KindDef {
  kind: string
  label: string
  emoji: string
  /**
   * Optional simple-icons / inline-brand key. When set, the admin
   * UI renders the official brand mark via <BrandIcon>; emoji is
   * used as fallback (and in compact list views where the SVG is
   * overkill).
   */
  iconKey?: string
  description: string
  /**
   * Fields the operator fills in. Order matches the form display.
   */
  fields: KindField[]
  /**
   * Fields whose presence in `config` should be rendered on the
   * channel card as a masked "token preview" line. First match wins.
   */
  tokenFields?: string[]
  /**
   * When true the channel needs a publicly-reachable webhook URL the
   * operator has to paste into the platform's admin console. The card
   * surfaces it via "Webhook URL" with a copy button.
   */
  webhookBased?: boolean
  /**
   * Optional after-create hint shown as a callout in the create
   * dialog. Useful for "remember to paste this URL into Feishu".
   */
  afterCreateHint?: string
}

export const KIND_DEFS: KindDef[] = [
  {
    kind: 'telegram',
    label: 'Telegram',
    emoji: '✈️',
    iconKey: 'telegram',
    description:
      'Bot via @BotFather. opendray long-polls getUpdates and sends via REST. Buttons + reply_to_message work natively.',
    tokenFields: ['bot_token'],
    fields: [
      {
        name: 'bot_token',
        label: 'Bot token',
        type: 'password',
        required: true,
        placeholder: '123456:ABC-DEF...',
        hint: 'From @BotFather. Stored in channel config; admin-only API.',
      },
      {
        name: 'chat_id',
        label: 'Default chat ID',
        type: 'text',
        placeholder: '42 (optional — used for outbound when no ReplyCtx)',
        optional: true,
      },
      {
        name: 'owner_user_ids',
        label: 'Owner Telegram user ID(s)',
        type: 'text',
        optional: true,
        placeholder: '123456789 (comma-separated for more than one)',
        hint: 'Only these numeric Telegram user IDs may drive sessions, run commands, or tap buttons; everyone else is ignored. Leave blank to allow anyone (not recommended for two-way chat). Get yours by DMing @userinfobot.',
      },
      {
        name: 'chat_enabled',
        label: 'Two-way chat (route messages into the session)',
        type: 'boolean',
        default: true,
        hint: 'When on, your messages are typed into the selected session and the agent replies here. Turn off for notifications only.',
      },
      {
        name: 'chat_typing',
        label: 'Show “typing…” while the agent works',
        type: 'boolean',
        default: true,
        hint: 'Displays a typing indicator until the agent’s reply settles.',
      },
      {
        name: 'reply_max_chars',
        label: 'Max reply length (characters)',
        type: 'text',
        optional: true,
        placeholder: '3500 (blank = 3500, 0 = unlimited)',
        hint: 'Caps how much of the agent’s reply is sent before it’s trimmed with a “…(truncated)” note. Blank uses the 3500 default (~one message); set 0 to send the whole reply, split across multiple messages.',
      },
    ],
  },
  {
    kind: 'slack',
    label: 'Slack',
    emoji: '💬',
    iconKey: 'slack',
    description:
      'Socket Mode — no public webhook needed. Requires both a bot OAuth token (xoxb-) and an app-level token (xapp-) with connections:write.',
    tokenFields: ['bot_token'],
    fields: [
      {
        name: 'bot_token',
        label: 'Bot token (xoxb-…)',
        type: 'password',
        required: true,
        placeholder: 'xoxb-...',
        hint: 'OAuth & Permissions → Bot User OAuth Token. Needs chat:write.',
      },
      {
        name: 'app_token',
        label: 'App-level token (xapp-…)',
        type: 'password',
        required: true,
        placeholder: 'xapp-...',
        hint: 'Settings → Basic Information → App-Level Tokens. Scope: connections:write.',
      },
      {
        name: 'channel_id',
        label: 'Default channel ID',
        type: 'text',
        placeholder: 'C0123ABC456 (optional)',
        optional: true,
      },
    ],
  },
  {
    kind: 'discord',
    label: 'Discord',
    emoji: '🎮',
    iconKey: 'discord',
    description:
      'Bot via Discord Developer Portal with MESSAGE CONTENT INTENT enabled. Connects to Gateway WS — no public URL required.',
    tokenFields: ['bot_token'],
    fields: [
      {
        name: 'bot_token',
        label: 'Bot token',
        type: 'password',
        required: true,
        placeholder: 'Bot token from Discord Developer Portal',
        hint: 'Application → Bot → Reset Token. Invite bot with send_messages + embed_links.',
      },
      {
        name: 'channel_id',
        label: 'Default channel ID',
        type: 'text',
        placeholder: '123456789012345678 (right-click channel → Copy ID)',
        optional: true,
      },
    ],
  },
  {
    kind: 'feishu',
    label: 'Feishu (飞书)',
    emoji: '🐦',
    iconKey: 'feishu',
    description:
      'App-level credentials. Uses event subscription webhook for inbound. Public webhook URL is generated below — paste it into the Feishu dev console.',
    tokenFields: ['app_secret'],
    webhookBased: true,
    afterCreateHint:
      'Open the webhook URL on the channel card and paste it into Feishu Open Platform → Event Subscriptions → Request URL.',
    fields: [
      {
        name: 'app_id',
        label: 'App ID',
        type: 'text',
        required: true,
        placeholder: 'cli_a1b2c3d4...',
      },
      {
        name: 'app_secret',
        label: 'App secret',
        type: 'password',
        required: true,
        placeholder: 'Application credential secret',
      },
      {
        name: 'verification_token',
        label: 'Verification token (optional)',
        type: 'password',
        optional: true,
        hint: 'From Event Subscriptions → Verification Token. When set, opendray rejects webhooks with a different token.',
      },
      {
        name: 'chat_id',
        label: 'Default chat ID (oc_…)',
        type: 'text',
        placeholder: 'oc_xxxxxxxxxx (optional)',
        optional: true,
      },
    ],
  },
  {
    kind: 'dingtalk',
    label: 'DingTalk (钉钉)',
    emoji: '📞',
    iconKey: 'dingtalk',
    description:
      'Custom group robot. Outbound only (text + markdown + actionCard). Group chat → Robots → Add → Sign mode → copy webhook + secret.',
    tokenFields: ['secret', 'webhook_url'],
    fields: [
      {
        name: 'webhook_url',
        label: 'Webhook URL',
        type: 'password',
        required: true,
        placeholder: 'https://oapi.dingtalk.com/robot/send?access_token=...',
      },
      {
        name: 'secret',
        label: 'Sign secret (optional)',
        type: 'password',
        optional: true,
        placeholder: 'SEC...',
        hint: 'When the robot is set to "Sign" security mode, copy the secret here. opendray adds the timestamp + sign query params automatically.',
      },
    ],
  },
  {
    kind: 'wecom',
    label: 'WeCom (企业微信)',
    emoji: '🏢',
    iconKey: 'wecom',
    description:
      'Group robot webhook. Outbound only (text + markdown). Group settings → Group robots → Add → copy webhook URL.',
    tokenFields: ['webhook_key', 'webhook_url'],
    fields: [
      {
        name: 'webhook_key',
        label: 'Webhook key',
        type: 'password',
        required: true,
        placeholder: 'The "key=" query value from the robot webhook URL',
        hint: 'Or paste the whole webhook URL into the field below — either is enough.',
      },
      {
        name: 'webhook_url',
        label: 'Or full webhook URL',
        type: 'password',
        optional: true,
        placeholder: 'https://qyapi.weixin.qq.com/cgi-bin/webhook/send?key=...',
      },
    ],
  },
  // Personal-WeChat (WxPusher) used to live here. Removed as a
  // create-flow option because it's outbound-only — operators
  // expecting a two-way channel had a frustrating discovery
  // experience. The server-side adapter stays so historical
  // channels keep working; only the New Channel entry point and
  // its tutorial are gone.
]

export function getKindDef(kind: string): KindDef | undefined {
  return KIND_DEFS.find((k) => k.kind === kind)
}

/**
 * Build the per-channel public webhook URL the platform's admin
 * console needs to call. Used for feishu / dingtalk / wecom (kinds
 * that opt-in via `webhookBased: true`).
 */
export function buildWebhookURL(channelID: string): string {
  const origin =
    typeof window !== 'undefined' ? window.location.origin : 'http://localhost:5173'
  return origin + '/api/v1/channels/' + channelID + '/webhook'
}
